docs/architecture/grpc-for-wcf-developers/types-of-rpc.md

```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
+        });
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        await _stream.RequestStream.CompleteAsync();
+        _stream.Dispose();
+    }
+}
+```
+
+Again, client streaming RPCs can be used for fire-and-forget messaging as shown above, but also for sending very large datasets to the server. The same caveat regarding performance applies: you should used `repeated` fields in regular messages for smaller datasets.
+
+## WCF full duplex services
+
+WCF duplex binding supports multiple one-way operations on both the service interface and the client callback interface, allowing ongoing conversations between client and server. gRPC supports something similar with Bi-directional Streaming RPCs, where both parameters are marked with the `stream` modifier.
+
+### chat.proto
+
+```protobuf
+service Chatter {
+    rpc Connect(stream IncomingMessage) returns (stream OutgoingMessage);
+}
+```
+
+### ChatterService.cs
+
+```csharp
+public class ChatterService : Chatter.ChatterBase
+{
+    private readonly IChatHub _hub;
+
+    public ChatterService(IChatHub hub)
+    {
+        _hub = hub;
+    }
+
+    public override async Task Connect(IAsyncStreamReader<MessageRequest> requestStream, IServerStreamWriter<MessageResponse> responseStream, ServerCallContext context)
+    {
+        _hub.MessageReceived += async (sender, args) =>
+            await responseStream.WriteAsync(new MessageResponse {Text = args.Message});
+
+        while (await requestStream.MoveNext(context.CancellationToken))
+        {
+            await _hub.SendAsync(requestStream.Current.Text);
+        }
+    }
+}
+```
+
+Here you can see that the implementation method receives both a request stream (`IAsyncStreamReader<MessageRequest>`) and a response stream (`IServerStreamWriter<MessageResponse>`), and can read and write messages until the connection is closed.
+
+### Chatter client
+
+```csharp
+public class Chat : IAsyncDisposable
+{
+    private readonly Chatter.ChatterClient _client;
+    private readonly AsyncDuplexStreamingCall<MessageRequest, MessageResponse> _stream;
+    private readonly CancellationTokenSource _cancellationTokenSource;
+    private readonly Task _readTask;
+
+    public Chat(Chatter.ChatterClient client)
+    {
+        _client = client;
+        _stream = _client.Connect();
+        _cancellationTokenSource = new CancellationTokenSource();
+        _readTask = ReadAsync(_cancellationTokenSource.Token);
+    }
+
+    public async Task SendAsync(string message)
+    {
+        await _stream.RequestStream.WriteAsync(new MessageRequest {Text = message});
+    }
+
+    private async Task ReadAsync(CancellationToken token)
+    {
+        while (await _stream.ResponseStream.MoveNext(token))
+        {
+            Console.WriteLine(_stream.ResponseStream.Current.Text);
+        }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        _cancellationTokenSource.Cancel();
```

Comment:
Is the cancellation token required? Completing the client request stream will cause the RequestStream.MoveNext on the server to return false, and the gRPC method to finish on the server. Then ResponseStream.MoveNext on the client will return false and the read task will complete.

It is a cleaner way to finish the gRPC call compared to cancellation.

```cs
        await _stream.RequestStream.CompleteAsync();
        await _readTask;
        _stream.Dispose();
```